### ISSUE TITLE
Add redirection functionality to recvfrom()

### ISSUE THREAD

**Issue by erikbor** [riboseinc/retrace#20]
This one is a bit tricky because we want to transmit ascii files based on a condition set by sendto(). As an example:
  1) a connect() is made to '192.168.1.1:80', which is then redirected to '10.1.1.1:8080'
  2) sendto() will receive one or two requests, e.g.:
     a) GET /
     b) GET /status
  3) recvfrom() should in case of 'sendto() a' send '/tmp/recvfrom-index.txt' and in case of 'sendto() b' send '/tmp/recvfrom-status.txt'
     A connect() to anything other than '192.168.1.1:80' will not have recvfrom() redirection.

**Comment by drystone** [comment]
I'll have a go at this. Just to clarify. If connect() is called with the 192 address, the ip is replaced with the 10... address. Then sendto() is essentially making http requests and instead of a response from 10... recvfrom() gets the contents of a file (for very specific urls)?

**Comment by erikbor** [comment]
Yes exactly, if there the "GET /urlx" is present in the config file then it needs to show the contents as set in the config file.

**Comment by riataman** [comment]
I have been thinking a bit about this one, I think this needs to look like this:

1) There's already code to redirect one address to another, that is done using the config file and one (or more) lines like this:

connect,127.0.0.1,8080,192.168.1.110,9090

This was the easy case!

2) To effectively implement recv redirection we would need a config file like this:

connect,127.0.0.1:80,"GET /",/home/test/response

Now, when the traced programs does:

connect("127.0.0.1", 80);

Internally we won't call connect at all! We would just return success and remember the file descriptor. Send in write/sendto calls we will assemble the TCP stream (remember that not necessarily the whole stream will come in a single call, so we need to keep track of this).

Once we detect a redirection match in the write/sendto stream, THEN we will make the file available for future read/readfrom calls (up to this point all calls to read/readfrom) would have no data, so they would either have to block or return 0 depending on the socket type.

That's the way I see we could implement this.

**Comment by jjr840430** [comment]
@ronaldtse @erikbor @drystone @riataman 
Can you please review it and let me know your opinions?

I think we should make more flexible redirecting solution.

To do it, we can make plug-in structure to customize each application protocols.
For example, we can make following structure.(It's just example)

// plug-in context structure declaration

typedef struct retrace_sock_plugin_ctx
{
        int (*p_init)();
        void (*p_finalize)();
        
        int (*p_connect_func)(int fd, const struct sockaddr *orig_addr, socklen_t orig_len,
                        sturct sockaddr *redirect_addr, socklent_t *redirect_len);

        int (*p_sendto_func)(int fd, const void *orig_buf, size_t len, void **new_buf, size_t *new_buf_s);
        int (*p_recvfrom_func)(int sockfd, void *buf, size_t len, int flags, struct sockaddr *src_addr, socklen_t *addrlen);

        void *persist_data;                     // this data will maintain information related the plugin context
} retrace_sock_plugin_ctx_t;

// sock.c in retrace

int RETRACE_IMPLEMENTATION(connect)(int fd, const struct sockaddr *address, socklen_t len)
{
        struct sockaddr *plugin_address;
        socklen_t plugin_len = sizeof(struct sockaddr);

        // call plugin function to get redirecting address
        p_connect_func(fd, address, len, &plugin_address, &plugin_len);

        return real_connect(fd, address, len);
}

int RETRACE_IMPLEMENTATION(sendto)(int sockfd, const void *buf, size_t len, int flags,
                      const struct sockaddr *dest_addr, socklen_t addrlen)
{
        void *new_buf;
        size_t new_buf_s;

        // call plugin function to send specific data
        p_sendto_func(fd, buf, len, &new_buf, &new_buf_s);

        return real_sendto(fd, new_buf, new_buf_s, flags, dest_addr, addrlen);
}

int RETRACE_IMPLEMENTATION(recvfrom)(int sockfd, void *buf, size_t len, int flags,
                        struct sockaddr *src_addr, socklen_t *addrlen)
{
        // call plugin function to recieve data
        p_recvfrom_func(fd, buf, len, flags, src_addr, addrlen);

        return real_recvfrom(fd, buf, len, flags, src_addr, addrlen);
}

// HTTP plugin context structure to manipluate HTTP data
struct http_plugin_data
{
        struct redirect_config *redirect_configs;
        struct recv_config *recv_config;
        struct sendto_config *sendto_config;
        ....
};

struct http_plugin_data http_ctx;

retrace_sock_plugin_ctx_t http_plugin_ctx =
{
        .p_init = http_init,
        .p_finalize = http_finalize,

        .p_connect = http_connect,
        .p_sendto = http_sendto,
        .p_recvfrom = http_recvfrom,

        .persist_data = (void *) &http_ctx,
};

int http_init()
{
        // read configuration from file
        // this will charge persist_data
}

void http_finalize()
{
        // finalize http plugin
        // this will free persist_data
}

int http_connect()
{
        // search redirection info from persist_data
}

int http_sendto()
{
        // if flag was marked, then process correspoding actions
}

int http_recvfrom()
{
        // search recvfrom match info into received data and mark flag into persist_data
}

**Comment by riataman** [comment]
Well, I was thinking towards something like this at the beginning, even allowing the plugins to be written in lua/python or something like this.

Then started thinking a bit more and there's no much benefit to it. There are a few uses cases that I can think of:

1) You want to redirect the connection from one IP to another. This is the easy case and is already implemented.
2) You want to fake 100% of the responses from a server. This is the option I put in above. The other option that someone doing this would have is doing 1) and then writing their own server to fake the responses, I don't think there would be much advantage introducing a plugin system to retrace overt just redirecting to a server of your own.
3) You want to fake a few bits and pieces of the stream. This is very, very difficult and you would probably need very protocol specific code for this one AND custom code. This is difficult because in some cases you might want to pass the stream from one side but fake the response, all while still reading the real response to discard it from the recv buffer (needs to basically know the whole state of the protocol), but sometimes you want to intercept the request AND don't pass it along and fake the response.

I think 2) is something that could be somewhat easily implemented (still need to reassemble the stream and keep track of it, so not that easy).

3) is something that is going to be super hard to implement and would probably be better left to some specialized tool (like rewriting http proxies).

**Comment by jjr840430** [comment]
@riataman Thanks for your reviews.

You are right. If we are going to fake HTTP traffics, then it's very hard to implement it.
This example is just ensure that we should use flexible and extensible plug-in structure to hook system calls.

Also if we need to build HTTP hooking solution, then we may use libraries for reassembling HTTP traffics.

For example, I think we can use this library.
http://libnids.sourceforge.net/

**Comment by riataman** [comment]
LibNIDS gives TCP stream reassembly from raw packages. Its useful when you only have access to raw packages, like when using PCAP to capture packages.

In retrace we already have the tcp stream assembled at the package level, we just need to keep track of the buffers as read/writes can come up in any size, not necessary in a whole buffer read.

Think of a client sending the string "test";

It can send it like this:

write (fd, "test", 4);

or it can send it like this:

write (fd, "t", 1);
write (fd, "e", 1);
write (fd, "s", 1);
write (fd, "t", 1);

or write can just send partial string, in which case the traced program needs to retry the send.

sendlen = write(fd, "test", 4)
if (sendlen < 4)
    write (fd, "test" + sendlen, 4 - sendlen)


so we need to be aware of all those sizes and keep track of them in our side until we have something meaningful that we can actually compare to.

**Comment by jjr840430** [comment]
You are right. Libnids is for raw packets. I confused something.

Then, I think we may think about following solution.
- hooking sendto system call(client => server)
A HTTP request may divided into several socket buffers. So we should keep buffers until a HTTP request is completed to be sent. At this case, we don't need to send fake buffers to server.
We can determine sendto buffer has '/GET' or '/GET status' after completed.

- hooking recvfrom system call(server => client)
The server will send HTTP response for above sending request.
Like as above case, we should keep buffers from server until a HTTP response is completed. At this case, we should not call real_recvfrom function.
After completed HTTP response, we should reassemble HTTP header and payload.

If past sendto buffer is /GET or /GET status, then we can modify HTTP header/payload and call real_recvfrom function.

**Comment by drystone** [comment]
@jjr840430 as I'd assigned this to myself (see top of page) and notified my intent to work on this issue via [ribose](https://www.ribose.com/#spaces/40b2063b-ccf5-484f-9fe9-fb7ffcafbedd/discussion/discussions/17798/messages/24568), I'm a little surprised to see you've been working on it too!

**Comment by jjr840430** [comment]
@drystone, Sorry if my working on this surprised you.
But I didn't complete this part, I just wanted to ensure that plugin structure may help you to solve this problem.
So I made just example plugin for HTTP socket functions.

**Comment by drystone** [comment]
@jjr840430 yes, I was hoping to. I've got a bit involved in bugfixing existing code the last couple of days!

**Comment by ronaldtse** [comment]
@drystone @jjr840430 any update on this issue? Hope we could complete this soon. Thanks!

**Comment by drystone** [comment]
Yes, sorry, I've effectively blocked this issue. I can't help thinking that the most flexible solution would be, as @riataman suggested, to set up a forwarding proxy to intercept the necessary urls and use the `connect` redirection that already exists. Technically though, I don't see any problem with redirecting simple (non ssl) HTTP requests. Config: `httpredirect,ip,port,contentdir`. The files in `contentdir` will need to be fully formed http responses including headers and any `Accept` header in the request will be ignored. I'll give it a go.

**Comment by jjr840430** [comment]
@ronaldtse I have written some simple plugin to implement this function.
Here is the link. https://github.com/riboseinc/retrace/tree/issue_20/plugin
I will try integrate it with recent retrace in next week.

**Comment by ronaldtse** [comment]
Shipped: redirection is a first-class config surface — `modify_in_param_str/int/arr` rewrite arguments pre-call, `addr_deny` gates connect targets, `modify_return_value_int` fabricates returns (docs/configuration.md, cookbook 06).
